korangar/src/loaders/animation/mod.rs

```diff
@@ -0,0 +1,474 @@
+            None => self.load(sprite_loader, action_loader, entity_filename, entity_type),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct FramePart {
+    pub animation_index: i32,
+    pub sprite_number: i32,
+    pub offset: Vector2<i32>,
+    pub upleft: Vector2<i32>, // Is only used for internal calculation
+    pub size: Vector2<i32>,
+    pub mirror: bool,
+    pub angle: f32,
+    pub color: Color,
+}
+
+#[derive(Clone)]
+pub struct Frame {
+    pub offset: Vector2<i32>,
+    pub upleft: Vector2<i32>,
+    pub size: Vector2<i32>,
+    pub remove_offset: Vector2<i32>, // Is only used for final shift
+    pub frameparts: Vec<FramePart>,
+}
+
+impl Frame {
+    // This function generate new rame
+    pub fn merge_frame_part(frames: &mut Vec<Frame>) -> Frame {
+        for frame in frames.iter_mut() {
+            // Finding the half size of the image
+            let half_size = (frame.size - Vector2::new(1, 1)) / 2;
+            frame.upleft = frame.offset - half_size;
+        }
+        // If there is no frame return an image with 1 pixel.
+        if frames.is_empty() {
+            let frame_part = FramePart {
+                animation_index: -1,
+                sprite_number: -1,
+                size: Vector2::new(1, 1),
+                upleft: Vector2::zero(),
+                offset: Vector2::zero(),
+                mirror: false,
+                angle: 0.0,
+                color: Color {
+                    red: 0.0,
+                    blue: 0.0,
+                    green: 0.0,
+                    alpha: 0.0,
+                },
+            };
+            let frame = Frame {
+                size: Vector2::new(1, 1),
+                upleft: Vector2::zero(),
+                offset: Vector2::zero(),
+                remove_offset: Vector2::zero(),
+                frameparts: vec![frame_part],
+            };
+            return frame;
+        }
+        // Find the upmost and leftmost coordinates
+        let upleft_x = frames.iter().min_by_key(|frame| frame.upleft.x).unwrap().upleft.x;
+        let upleft_y = frames.iter().min_by_key(|frame| frame.upleft.y).unwrap().upleft.y;
+
+        // Find the downmost and rightmost coordinates
+        let frame_x = frames.iter().max_by_key(|frame| frame.upleft.x + frame.size.x as i32).unwrap();
+        let frame_y = frames.iter().max_by_key(|frame| frame.upleft.y + frame.size.y as i32).unwrap();
+
+        // Calculate the new rectangle that is formed
+        let mut new_width = frame_x.upleft.x + frame_x.size.x as i32;
+        let mut new_height = frame_y.upleft.y + frame_y.size.y as i32;
+        new_width -= upleft_x;
+        new_height -= upleft_y;
+
+        let mut new_frameparts = Vec::<FramePart>::new();
+        for index in 0..frames.len() {
+            new_frameparts.append(&mut frames[index].frameparts);
+        }
+
+        /* As the origin is (0,0), you get the upleft point of the retangle and

```

Comment:
We normally use line comments for multi line comments too. Here are also some typos.

Reply:
Resolved